runtime/src/main/java/org/corfudb/runtime/collections/streaming/DeltaStream.java

```diff
@@ -0,0 +1,184 @@
+package org.corfudb.runtime.collections.streaming;
+
+import com.google.common.base.Preconditions;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections.Buffer;
+import org.apache.commons.collections.BufferUtils;
+import org.apache.commons.collections.buffer.BoundedFifoBuffer;
+import org.corfudb.common.metrics.micrometer.MicroMeterUtils;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.AddressSpaceView;
+import org.corfudb.runtime.view.ReadOptions;
+import org.corfudb.runtime.view.stream.StreamAddressSpace;
+
+import java.time.Duration;
+import java.util.UUID;
+
+/**
+ * Unlike the {@link org.corfudb.runtime.view.stream.IStreamView} this stream implementation only tracks updates
+ * that haven't been checkpointed.
+ */
+@Slf4j
+public class DeltaStream {
+
+    private final AddressSpaceView addressSpaceView;
+
+    @Getter
+    private final UUID streamId;
+
+    /**
+     * Last Address consumed/read by this stream
+     */
+    private volatile long lastAddressRead;
+
+    /**
+     * Max seen address for this stream from the last refresh/sync
+     */
+    @Getter
+    private volatile long maxAddressSeen;
+
+    /**
+     * The last trim mark that has been observed
+     */
+    private volatile long trimMark;
+
+    /**
+     * The size of this stream's buffer (i.e., pending addresses to read)
+     */
+    private final int bufferSize;
+
+    /**
+     * A circular buffer that holds the queued address to read
+     */
+    private final Buffer addressesToRead;
+
+    private final ReadOptions options = ReadOptions
+            .builder()
+            .clientCacheable(false)
+            .ignoreTrim(false)
+            .waitForHole(true)
+            .serverCacheable(false)
+            .build();
+
+    public DeltaStream(AddressSpaceView addressSpaceView, UUID streamId, long lastAddressRead, int bufferSize) {
+        this.addressSpaceView = addressSpaceView;
+        Preconditions.checkArgument(Address.NON_ADDRESS == -1);
```

Comment:
looks weird, what's the motivation for comparing two constants?

Reply:
It does look weird, but if you look at the subsequent line, there's logic that depends on the specific value (i.e., -1) and I don't want to use -1 directly. This just makes sure that if NON_ADDRESS changes, then these tests would fail.

```java

Preconditions.checkArgument(Address.NON_ADDRESS == -1);
Preconditions.checkArgument(lastAddressRead >= Address.NON_ADDRESS,
                "lastAddressRead %s must be >= %s", lastAddressRead, Address.NON_ADDRESS);

```

Reply:
Now, that's a weird motivation for that also. 
If you want NON_ADDRESS to be always -1, then write a unit test that checks NON_ADDRESS == -1.